lxd/instance/drivers/driver_qemu.go

```diff
@@ -3655,7 +3657,7 @@ func (d *qemu) deviceAdd(deviceName string, rawConfig deviceConfig.Device, insta
 }
 
 func (d *qemu) deviceRemove(deviceName string, rawConfig deviceConfig.Device, instanceRunning bool) error {
-	logger := logging.AddContext(logger.Log, log.Ctx{"device": deviceName, "type": rawConfig["type"], "project": d.Project(), "instance": d.Name()})
+	logger := logging.AddContext(d.common.logger, log.Ctx{"device": deviceName, "type": rawConfig["type"], "project": d.Project(), "instance": d.Name()})
```

Comment:
@kevtheappdev so for example here, you've correctly create a new local `logger`, but use `d.logger` as the source logger and there's no need to add `instance` and `project` again as they will be inherited.

Reply:
Thanks for the response! In instantiating the logger instance on `d` we are only setting the `instance` and `name` fields: `logging.AddContext(logger.Log, log.Ctx{"instanceType": args.Type.String, "instanceName": args.Name})`.
Just to clarify, are we to exclude fields already included in the context we provide to the logger instance we create on `d`? Or any fields that are included in `d`?

Reply:
So as an example take a look at the storage package and the storage pool workflow:

1. We load (and initialise) a storage pool "object" via `GetPoolByName` which internally inherits the global logger using `b.logger = logging.AddContext(logger.Log,...)` to setup a new internal logger with context info about the pool. https://github.com/lxc/lxd/blob/850dd0a5adf4e26f4faff5a4580e9492a32eb277/lxd/storage/load.go#L153
2. Then for each function inside the storage pool it sets up a new local logger using the object's internal logger as the basis (and inheriting its context) using `logger := logging.AddContext(b.logger,...)` https://github.com/lxc/lxd/blob/master/lxd/storage/backend_lxd.go#L521-L523
3. This then allows any logging call inside the function to just do `logger.Error("some message")` and every log message will contain *both* the function specific conxtext *and* the storage pool specific context (which is really useful, but avoids having to repeatedly define the same variables in the context over and over).
4. Further more if you need to add a one-time context to a log message (for example an error message) then you can just do `logger.Errorf("Failed doing something", log.Ctx{"err": err})` and it will *add* the "err" variable to the wider inherited context.